common/thorhelper/roxiehelper.cpp

```diff
@@ -411,8 +413,872 @@ bool CRHLimitedCompareHelper::getGroup(OwnedRowArray &group, const void *left)
+                break;
+            builder.appendEOG();
+        }
+        builder.appendOwn(nextrec);
+    }
+}
+
+
+//=========================================================================================
+
+// Ability to read an input stream and group and/or sort it on-the-fly
+
+using roxiemem::OwnedConstRoxieRow;
+
+class GroupedInputReader : public CInterfaceOf<IGroupedInput>
+{
+protected:
+    bool firstRead;
+    bool eof;
+    bool endGroupPending;
+    OwnedConstRoxieRow next;
+
+    IInputBase *input;
+    const ICompare *compare;
+public:
+    GroupedInputReader(IInputBase *_input, const ICompare *_compare)
+    : input(_input), compare(_compare)
+    {
+        firstRead = false;
+        eof = false;
+        endGroupPending = false;
+    }
+
+    virtual IOutputMetaData * queryOutputMeta() const
+    {
+        return input->queryOutputMeta();
+    }
+
+    virtual const void *nextInGroup()
+    {
+        if (!firstRead)
+        {
+            firstRead = true;
+            next.setown(input->nextInGroup());
+        }
+
+        if (eof || endGroupPending)
+        {
+            endGroupPending = false;
+            return NULL;
+        }
+
+        OwnedConstRoxieRow prev(next.getClear());
+        next.setown(input->nextUngrouped());  // skip incoming grouping if present
+
+        if (next)
+        {
+            assertex(prev);  // If this fails, you have an initial empty group. That is not legal.
+            if (compare && compare->docompare(prev, next) != 0) // MORE - could assert >=0, as input is supposed to be sorted
+                endGroupPending = true;
+        }
+        else
+            eof = true;
+        return prev.getClear();
+    }
+};
+
+//=========================================================================================
+
+class DegroupedInputReader : implements IInputBase
+{
+private:
+    IInputBase *input;
+public:
+    DegroupedInputReader(IInputBase *_input) : input(_input)
+    {
+    }
+    virtual IOutputMetaData *queryOutputMeta() const
+    {
+        return input->queryOutputMeta();
+    }
+    virtual const void *nextInGroup()
+    {
+        return input->nextUngrouped();
+    }
+};
+
+class SortedGroupedInputReader : public GroupedInputReader
+{
+protected:
+    DegroupedInputReader degroupedInput;
+    Owned<ISortAlgorithm> sorter;
+public:
+    SortedGroupedInputReader(IInputBase *_input, const ICompare *_compare, ISortAlgorithm *_sorter)
+      : GroupedInputReader(_input, _compare), degroupedInput(_input), sorter(_sorter)
+    {
+        sorter->reset();
+    }
+
+    virtual const void *nextInGroup()
+    {
+        if (!firstRead)
+        {
+            firstRead = true;
+            sorter->prepare(&degroupedInput);
+            next.setown(sorter->next());
+        }
+
+        if (eof || endGroupPending)
+        {
+            endGroupPending = false;
+            return NULL;
+        }
+
+        OwnedConstRoxieRow prev(next.getClear());
```

Comment:
the code here and following could be in an inline function and shared with the base class.  Not sure it is worth it though.